src/lib.rs

```diff
@@ -176,457 +138,684 @@ impl FromStr for ZKConnectString {
-    /// Arc/Mutex because the channel must be used by various threads in various
-    /// methods, and, unlike with the Sender end, we can't handle Recevier
-    /// ownership by cloning it.
-    event_rx: Arc<Mutex<Receiver<ResolverEvent>>>,
-    /// The last error that occurred. Protected by an Arc/Mutex because it is
-    /// accessed by multiple internal threads.
-    error: Arc<Mutex<Option<String>>>,
+    ///
     /// The key representation of the last backend sent to the cueball
-    /// connection pool. Protected by an Arc/Mutex because it is
-    /// accessed by the background thread in run(), but must be persistently
-    /// stored across starts/stops in this struct.
+    /// connection pool. Persists across multiple calls to run().
+    ///
     last_backend: Arc<Mutex<Option<BackendKey>>>,
-    /// Indicates if the resolver is running. Protected by an Arc/Mutex because
-    /// it is accessed by multiple internal threads.
-    running: Arc<Mutex<bool>>,
+    ///
+    /// Indicates whether or not the resolver is running. This is slightly
+    /// superfluous (this field is `true` for exactly the duration of each
+    /// call to run(), and false otherwise), but could be useful if the caller
+    /// wants to check if the resolver is running for some reason.
+    ///
+    is_running: bool,
+    ///
+    /// The ManateePrimaryResolver's root log
+    ///
+    log: Logger
 }
 
 impl ManateePrimaryResolver {
-
-    /// Creates a new ManateePrimaryResolver instance
+    ///
+    /// Creates a new ManateePrimaryResolver instance.
     ///
     /// # Arguments
     ///
     /// * `connect_string` - a comma-separated list of the zookeeper instances
     ///   in the cluster
     /// * `path` - The path to the root node in zookeeper for the shard we're
-    //    watching
+    ///    watching
+    ///
     pub fn new(
-        connect_string: ZKConnectString,
+        connect_string: ZkConnectString,
         path: String
     ) -> Self
     {
-        // TODO change 'dean' to 'state' once you're done testing
-        let cluster_state_path = [&path, "/dean"].concat();
-        let (tx, rx) = channel();
+        let cluster_state_path = [&path, "/state"].concat();
 
         ManateePrimaryResolver {
-            connect_string,
-            cluster_state_path,
-            pool_tx: None,
-            event_tx: tx,
-            event_rx: Arc::new(Mutex::new(rx)),
-            error: Arc::new(Mutex::new(None)),
+            connect_string: connect_string.clone(),
+            cluster_state_path: cluster_state_path.clone(),
             last_backend: Arc::new(Mutex::new(None)),
-            running: Arc::new(Mutex::new(false)),
+            is_running: false,
+            log: Logger::root(
+                Mutex::new(LevelFilter::new(
+                    slog_bunyan::with_name(crate_name!(),
+                        std::io::stdout()).build(),
+                    //
+                    // TODO replace with user-configurable log level? How?
```

Comment:
I think as a library the right call is to optionally allow a `Logger` to be passed in as a parameter to `new` with the log level for the encompassing application set and if it's not provided then to use a reasonable default like logging to STDOUT. This is what I did for `rust-cueball` and what the slog docs point to [here](https://github.com/slog-rs/example-lib/blob/master/lib.rs) as the way to deal with logging in libraries.

Reply:
Done -- copied code from rust-cueball. Note that clippy complained about my use of `unwrap_or` with a function call and told me to use `unwrap_or_else` instead.